testsuite/libra-fuzzer/src/fuzz_targets/secure_storage_vault.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{corpus_from_strategy, fuzz_data_to_value, FuzzTargetImpl};
+use libra_proptest_helpers::ValueGenerator;
+use libra_vault_client::{
+    fuzzing::{arb_generic_response, arb_policy_list_response, arb_secret_read_response},
+    process_generic_response, process_policy_list_response, process_secret_read_response,
+};
+use proptest::arbitrary::any;
+
+#[derive(Clone, Debug, Default)]
+pub struct VaultGenericResponse;
+
+/// This implementation will fuzz process_generic_response(): the method used by the vault
+/// client to process generic responses from the vault backend.
+impl FuzzTargetImpl for VaultGenericResponse {
+    fn description(&self) -> &'static str {
+        "Secure storage vault: process_generic_response()"
+    }
+
+    fn generate(&self, _idx: usize, _gen: &mut ValueGenerator) -> Option<Vec<u8>> {
+        Some(corpus_from_strategy(arb_generic_response()))
+    }
+
+    fn fuzz(&self, data: &[u8]) {
+        let input = fuzz_data_to_value(data, arb_generic_response());
+        let _ = process_generic_response(input);
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct VaultPolicyListResponse;
+
+/// This implementation will fuzz process_policy_list_response(): the method used by the vault
+/// client to process policy lists from the vault backend.
+impl FuzzTargetImpl for VaultPolicyListResponse {
+    fn description(&self) -> &'static str {
+        "Secure storage vault: process_policy_list_response()"
+    }
+
+    fn generate(&self, _idx: usize, _gen: &mut ValueGenerator) -> Option<Vec<u8>> {
+        Some(corpus_from_strategy(arb_policy_list_response()))
+    }
+
+    fn fuzz(&self, data: &[u8]) {
+        let input = fuzz_data_to_value(data, arb_policy_list_response());
+        let _ = process_policy_list_response(input);
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct VaultSecretReadResponse;
+
+/// This implementation will fuzz process_secret_read_response(): the method used by the vault
+/// client to process secrets read from the vault backend.
+impl FuzzTargetImpl for VaultSecretReadResponse {
+    fn description(&self) -> &'static str {
+        "Secure storage vault: process_secret_read_response()"
+    }
+
+    fn generate(&self, _idx: usize, _gen: &mut ValueGenerator) -> Option<Vec<u8>> {
+        Some(corpus_from_strategy(arb_secret_read_response()))
+    }
+
+    fn fuzz(&self, data: &[u8]) {
+        let secret = fuzz_data_to_value(data, any::<String>());
+        let key = fuzz_data_to_value(data, any::<String>());
+        let response = fuzz_data_to_value(data, arb_secret_read_response());
```

Comment:
OK you actually cannot do this : o and maybe I should refactor fuzz_data_to_value to make it clearer. The reason why is that you're going to use the same data to produce secret, key, and response. Instead you want to use different parts of data to do this. I would suggest doing something like this:

```
let (secret, key, response) = fuzz_data_to_value(data, some_strategy());
```

Reply:
does secret and key have to be fuzzed? In other words can they be fixed?

I'm not sure how proptest generates String, but I'm guessing it might have some size limitation (cc @sunshowers). I would recommend inverting the order of the strategy so that the lengths of secret & key don't affect the fuzzing of response (if that makes sense):

```
let (response, secret, key) = fuzz_data_to_value(data, some_strategy());
```

Reply:
Aah, gotcha! Fixed, thanks! 😄 

1. I think it's better if secret and key are fuzzed, for completeness sake.. Just to be clear though (for my understanding), technically, there is nothing stopping someone from taking the approach of calling 'fuzz_data_to_value()' multiple times, right?.. It's more just that it's not ideal a single change in the input data should change all 3 values.. rather we'd like a single change in the the input data to only change 1 of the values at a time.

2. Regarding the ordering, that's an open question to me, too. My feeling is that if there is some limit to the input data, then this could be a problem. Although, I'm interested to know what happens in these cases? Would it just repeat the input data as the source of randomness? Anyway, I've re-ordered them as suggested! 😄

Reply:
2 -> yeah you're right, you can call it multiple times but that's probably not what you want to do

Reply:
3 -> if proptest runs out of random bytes it'll use 0s to fill in the rest